src/services/Login.ts

```diff
@@ -0,0 +1,154 @@
+import * as API from './Api';
+
+import { AuthInfo, AuthStrategy, AuthResult } from '../types/Auth';
+import { LoginSession } from '../store/Store';
+import { TimeInMilliseconds } from '../types/Common';
+
+// We just mock the types for state for now, to avoid injecting more
+// dependencies on the service.
+type Dispatch = any;
+type AppState = any;
+
+export interface LoginResult {
+  status: AuthResult;
+  session?: LoginSession;
+  error?: any;
+}
+
+interface LoginStrategy<T = {}> {
+  prepare: (info: AuthInfo) => Promise<AuthResult>;
+  perform: (request: DispatchRequest<T>) => Promise<LoginResult>;
+}
+
+interface DispatchRequest<T> {
+  dispatch: Dispatch;
+  getState: () => AppState;
+  data: T;
+}
+
+type NullDispatch = DispatchRequest<any>;
+
+class AnonymousLogin implements LoginStrategy {
+  public async prepare(info: AuthInfo) {
+    return AuthResult.CONTINUE;
+  }
+
+  public async perform(_request: NullDispatch): Promise<LoginResult> {
+    const session: LoginSession = (await API.login()).data;
+
+    return {
+      status: AuthResult.SUCCESS,
+      session: session
+    };
+  }
+}
+
+interface WebLoginData {
+  username: string;
+  password: string;
+}
+
+class WebLogin implements LoginStrategy<WebLoginData> {
+  public async prepare(info: AuthInfo) {
+    return AuthResult.CONTINUE;
+  }
+
+  public async perform(request: DispatchRequest<WebLoginData>): Promise<LoginResult> {
+    const session = (await API.login(request.data)).data;
+
+    return {
+      status: AuthResult.SUCCESS,
+      session: session
+    };
+  }
+}
+
+class OpenshiftLogin implements LoginStrategy<any> {
+  public async prepare(info: AuthInfo) {
+    if (info.authorizationEndpoint === undefined) {
+      return AuthResult.FAILURE;
+    }
+
+    if (window.location.hash.startsWith('#access_token')) {
+      return AuthResult.CONTINUE;
+    } else {
+      window.location.href = info.authorizationEndpoint!;
+
+      return AuthResult.HOLD;
+    }
+  }
+
+  public async perform(_request: NullDispatch): Promise<LoginResult> {
+    const session = (await API.checkOpenshiftAuth(window.location.hash.substring(1))).data;
+
+    return {
+      status: AuthResult.SUCCESS,
+      session: session
```

Comment:
You can use the ES6 shorthand instead of repeating:
```suggestion
      session
```